File: pipeline/geo/geocode_appellations.py
"""
Geocodes all appellations that don't yet have a geographic_boundaries row.
Uses OpenStreetMap Nominatim (free, 1 req/sec rate limit).
Inserts centroid-only rows with confidence='geocoded', source='nominatim'.

Usage: python -m pipeline.geo.geocode_appellations [--dry-run] [--limit N] [--verbose]
"""

import sys
import re
import time
import argparse
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

import httpx
from pipeline.lib.db import get_supabase

NOMINATIM_URL = "https://nominatim.openstreetmap.org/search"
RATE_LIMIT_MS = 1100
USER_AGENT = "Loam Wine Intelligence (loam.onrender.com)"

DESIGNATION_PREFIXES = [
    "AOC ", "AOP ", "DOC ", "DOCG ", "DO ", "AVA ", "IGP ", "IGT ",
    "VdP ", "GI ", "IG ", "PDO ", "PGI ", "DOP ", "WO ",
]

PLACE_EXTRACTORS = [
    re.compile(r"\bdi\s+(.+)$", re.IGNORECASE),
    re.compile(r"\bdel\s+(.+)$", re.IGNORECASE),
    re.compile(r"\bdella\s+(.+)$", re.IGNORECASE),
    re.compile(r"\bdelle\s+(.+)$", re.IGNORECASE),
    re.compile(r"\bdei\s+(.+)$", re.IGNORECASE),
    re.compile(r"\bd'\s*(.+)$", re.IGNORECASE),
    re.compile(r"\bde\s+(.+)$", re.IGNORECASE),
]

QUALITY_SUFFIXES = [
    "Superiore", "Riserva", "Classico", "Gran Selezione", "Passito",
    "Spumante", "Frizzante", "Liquoroso", "Novello",
]


def clean_appellation_name(name: str) -> str:
    cleaned = name
    for prefix in DESIGNATION_PREFIXES:
        if cleaned.startswith(prefix):
            cleaned = cleaned[len(prefix):]
    return cleaned.strip()


def strip_quality_suffix(name: str) -> str:
    cleaned = name
    for suffix in QUALITY_SUFFIXES:
        if cleaned.endswith(f" {suffix}"):
            cleaned = cleaned[:-(len(suffix) + 1)]
    return cleaned


def extract_place_names(name: str) -> list[str]:
    places = []
    seen = set()
    for variant in [name, strip_quality_suffix(name)]:
        for regex in PLACE_EXTRACTORS:
            match = regex.search(variant)
            if match:
                place = match.group(1).strip()
                if place not in seen:
                    seen.add(place)
                    places.append(place)
                break
    return places


def build_search_queries(appellation: str, region: str, country: str) -> list[str]:
    clean_name = clean_appellation_name(appellation)
    stripped_name = strip_quality_suffix(clean_name)
    place_names = extract_place_names(clean_name)
    queries = []
    seen = set()

    def add(q: str):
        if q not in seen:
            seen.add(q)
            queries.append(q)

    add(f"{clean_name}, {region}, {country}")
    if stripped_name != clean_name:
        add(f"{stripped_name}, {region}, {country}")
    add(f"{clean_name} wine region, {country}")
    for place in place_names:
        add(f"{place}, {region}, {country}")
        add(f"{place}, {country}")
    if not place_names:
        add(f"{clean_name}, {country}")
    if clean_name.lower() != region.lower():
        add(f"{region} wine region, {country}")
        add(f"{region}, {country}")
    return queries


def geocode_query(client: httpx.Client, query: str) -> dict | None:
    params = {"q": query, "format": "json", "limit": "1", "addressdetails": "0"}
    resp = client.get(NOMINATIM_URL, params=params, headers={
        "User-Agent": USER_AGENT, "Accept-Language": "en",
    })
    resp.raise_for_status()
    data = resp.json()
    if data:
        return {
            "lat": float(data[0]["lat"]),
            "lng": float(data[0]["lon"]),
            "display_name": data[0]["display_name"],
            "osm_type": data[0].get("osm_type"),
            "osm_id": data[0].get("osm_id"),
        }
    return None


def geocode_appellation(client: httpx.Client, appellation: str, region: str,
                        country: str, verbose: bool = False) -> dict | None:
    queries = build_search_queries(appellation, region, country)
    region_fallback_query = f"{region} wine region, {country}"

    for query in queries:
        time.sleep(RATE_LIMIT_MS / 1000)
        if verbose:
            print(f"     trying: \"{query}\" ... ", end="", flush=True)
        result = geocode_query(client, query)
        if result:
            if verbose:
                print("ok")
            is_region_fallback = query == region_fallback_query
            return {**result, "query": query, "regionFallback": is_region_fallback}
        if verbose:
            print("no result")
    return None


def main():
    parser = argparse.ArgumentParser(description="Geocode appellations via Nominatim")
    parser.add_argument("--dry-run", action="store_true", help="Preview only, no DB writes")
    parser.add_argument("--limit", type=int, default=None, help="Max appellations to process")
    parser.add_argument("--verbose", action="store_true", help="Show each query attempt")
    args = parser.parse_args()

    sb = get_supabase()
    print(f"Geocoding appellations{' (DRY RUN)' if args.dry_run else ''}")
    print(f"   Rate limit: {RATE_LIMIT_MS}ms per request\n")

    # Fetch appellations with region/country via join
    all_apps = []
    offset = 0
    while True:
        result = sb.table("appellations").select(
            "id, name, designation_type, regions!inner(name, countries!inner(name))"
        ).order("name").range(offset, offset + 999).execute()
        all_apps.extend(result.data)
        if len(result.data) < 1000:
            break
        offset += 1000

    # Check which already have boundaries
    existing_ids = set()
    offset = 0
    while True:
        result = sb.table("geographic_boundaries").select("appellation_id").not_(
            "appellation_id", "is", "null"
        ).range(offset, offset + 999).execute()
        for row in result.data:
            existing_ids.add(row["appellation_id"])
        if len(result.data) < 1000:
            break
        offset += 1000

    to_geocode = [a for a in all_apps if a["id"] not in existing_ids]

    print(f"   Total appellations: {len(all_apps)}")
    print(f"   Already geocoded: {len(existing_ids)}")
    print(f"   To geocode: {len(to_geocode)}")
    if args.limit:
        print(f"   Limit: {args.limit}")
    print()

    batch = to_geocode[:args.limit] if args.limit else to_geocode
    success = 0
    region_fallbacks = 0
    failed = 0
    failures = []

    client = httpx.Client(timeout=30)

    for i, a in enumerate(batch):
        region_name = a["regions"]["name"]
        country_name = a["regions"]["countries"]["name"]
        progress = f"[{i + 1}/{len(batch)}]"

        try:
            result = geocode_appellation(client, a["name"], region_name, country_name, args.verbose)

            if result:
                fallback_tag = " [region fallback]" if result["regionFallback"] else ""
                if result["regionFallback"]:
                    region_fallbacks += 1
                print(f"{progress} ok {a['name']} ({region_name}, {country_name}) -> "
                      f"{result['lat']:.4f}, {result['lng']:.4f}{fallback_tag}")

                if not args.dry_run:
                    point_wkt = f"SRID=4326;POINT({result['lng']} {result['lat']})"
                    source_id = (f"{result['osm_type']}/{result['osm_id']}"
                                 if result.get("osm_id") else None)

                    insert_result = sb.table("geographic_boundaries").insert({
                        "appellation_id": a["id"],
                        "centroid": point_wkt,
                        "boundary_confidence": "geocoded",
                        "boundary_source": "nominatim",
                        "boundary_source_id": source_id,
                    }).execute()

                    if hasattr(insert_result, "error") and insert_result.error:
                        print(f"   Insert failed: {insert_result.error}")
                        failed += 1
                        failures.append({
                            "name": a["name"], "region": region_name,
                            "country": country_name, "error": str(insert_result.error),
                        })
                        continue
                success += 1
            else:
                print(f"{progress} FAIL {a['name']} ({region_name}, {country_name}) -- no results")
                failed += 1
                failures.append({
                    "name": a["name"], "region": region_name,
                    "country": country_name, "error": "no geocode results",
                })
        except Exception as err:
            print(f"{progress} FAIL {a['name']} -- {err}")
            failed += 1
            failures.append({
                "name": a["name"], "region": region_name,
                "country": country_name, "error": str(err),
            })

    client.close()

    print(f"\n--- Summary ---")
    print(f"Geocoded:          {success}")
    print(f"  Region fallback: {region_fallbacks}")
    print(f"Failed:            {failed}")
    if failures:
        print(f"\nFailed appellations:")
        for f in failures:
            print(f"  - {f['name']} ({f['region']}, {f['country']}): {f['error']}")


if __name__ == "__main__":
    main()

File: scripts_archive/node/root/geocode_appellations.mjs
/**
 * geocode_appellations.mjs
 *
 * Geocodes all appellations that don't yet have a geographic_boundaries row.
 * Uses OpenStreetMap Nominatim (free, 1 req/sec rate limit).
 * Inserts centroid-only rows with confidence='geocoded', source='nominatim'.
 *
 * Usage: node geocode_appellations.mjs [--dry-run] [--limit N]
 */

import { createClient } from '@supabase/supabase-js'

const SUPABASE_URL = 'https://vgbppjhmvbggfjztzobl.supabase.co'
const SUPABASE_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InZnYnBwamhtdmJnZ2ZqenR6b2JsIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NzI1ODU1NDYsImV4cCI6MjA4ODE2MTU0Nn0.KHZiqk6B7XYDnkFcDNJtMIKoT-hf7s8MGkmpOsjgVDk'
const supabase = createClient(SUPABASE_URL, SUPABASE_KEY)

const NOMINATIM_URL = 'https://nominatim.openstreetmap.org/search'
const RATE_LIMIT_MS = 1100 // slightly over 1 second to be safe

const args = process.argv.slice(2)
const DRY_RUN = args.includes('--dry-run')
const limitIdx = args.indexOf('--limit')
const LIMIT = limitIdx >= 0 ? parseInt(args[limitIdx + 1]) : Infinity

const sleep = ms => new Promise(r => setTimeout(r, ms))

// Wine-specific designation prefixes to strip for cleaner geocoding
const DESIGNATION_PREFIXES = [
  'AOC ', 'AOP ', 'DOC ', 'DOCG ', 'DO ', 'AVA ', 'IGP ', 'IGT ',
  'VdP ', 'GI ', 'IG ', 'PDO ', 'PGI ', 'DOP ', 'WO '
]

function cleanAppellationName(name) {
  let cleaned = name
  for (const prefix of DESIGNATION_PREFIXES) {
    if (cleaned.startsWith(prefix)) {
      cleaned = cleaned.slice(prefix.length)
    }
  }
  return cleaned.trim()
}

// Common suffixes in Italian/Spanish DOC names that follow the place name
// e.g., "Aglianico del Vulture" → place is "Vulture"
// e.g., "Brunello di Montalcino" → place is "Montalcino"
// e.g., "Vino Nobile di Montepulciano" → place is "Montepulciano"
const PLACE_EXTRACTORS = [
  /\bdi\s+(.+)$/i,          // "Brunello di Montalcino" → "Montalcino"
  /\bdel\s+(.+)$/i,         // "Aglianico del Vulture" → "Vulture"
  /\bdella\s+(.+)$/i,       // "Vernaccia della Val Tidone" → "Val Tidone"
  /\bdelle\s+(.+)$/i,       // "Colli delle Marche" → "Marche"
  /\bdei\s+(.+)$/i,         // "Colli dei Bolognesi" → "Bolognesi"
  /\bd'\s*(.+)$/i,          // "Montepulciano d'Abruzzo" → "Abruzzo"
  /\bde\s+(.+)$/i,          // "Ribera de Duero" → "Duero" (less useful, but region helps)
]

// Quality tier suffixes commonly appended to Italian DOC/DOCG names
const QUALITY_SUFFIXES = [
  'Superiore', 'Riserva', 'Classico', 'Gran Selezione', 'Passito',
  'Spumante', 'Frizzante', 'Liquoroso', 'Novello'
]

/**
 * Strip quality tier suffixes: "Aglianico del Vulture Superiore" → "Aglianico del Vulture"
 */
function stripQualitySuffix(name) {
  let cleaned = name
  for (const suffix of QUALITY_SUFFIXES) {
    if (cleaned.endsWith(` ${suffix}`)) {
      cleaned = cleaned.slice(0, -(suffix.length + 1))
    }
  }
  return cleaned
}

/**
 * Extract all candidate place names from a wine appellation name.
 * Returns unique place names from both the original and suffix-stripped versions.
 * e.g., "Aglianico del Vulture Superiore" → ["Vulture Superiore", "Vulture"]
 */
function extractPlaceNames(name) {
  const places = []
  const seen = new Set()
  for (const variant of [name, stripQualitySuffix(name)]) {
    for (const regex of PLACE_EXTRACTORS) {
      const match = variant.match(regex)
      if (match) {
        const place = match[1].trim()
        if (!seen.has(place)) { seen.add(place); places.push(place) }
        break // only first regex match per variant
      }
    }
  }
  return places
}

/**
 * Build search queries in priority order — try the most specific first,
 * fall back to broader queries.
 */
function buildSearchQueries(appellation, region, country) {
  const cleanName = cleanAppellationName(appellation)
  const strippedName = stripQualitySuffix(cleanName)
  const placeNames = extractPlaceNames(cleanName)
  const queries = []
  const seen = new Set()

  function add(q) {
    if (!seen.has(q)) { seen.add(q); queries.push(q) }
  }

  // 1. Most specific: "appellation, region, country"
  add(`${cleanName}, ${region}, ${country}`)

  // 2. If quality suffix was stripped, try the base name
  if (strippedName !== cleanName) {
    add(`${strippedName}, ${region}, ${country}`)
  }

  // 3. "appellation wine region, country"
  add(`${cleanName} wine region, ${country}`)

  // 4. Extracted place names (e.g., "Vulture Superiore" then "Vulture")
  for (const place of placeNames) {
    add(`${place}, ${region}, ${country}`)
    add(`${place}, ${country}`)
  }

  // 5. Broader: just "appellation, country"
  if (placeNames.length === 0) {
    add(`${cleanName}, ${country}`)
  }

  // 6. Region fallback — at minimum get region-level accuracy
  if (cleanName.toLowerCase() !== region.toLowerCase()) {
    add(`${region} wine region, ${country}`)
    add(`${region}, ${country}`)
  }

  return queries
}

async function geocodeQuery(query) {
  const url = new URL(NOMINATIM_URL)
  url.searchParams.set('q', query)
  url.searchParams.set('format', 'json')
  url.searchParams.set('limit', '1')
  url.searchParams.set('addressdetails', '0')

  const res = await fetch(url, {
    headers: {
      'User-Agent': 'Loam Wine Intelligence (loam.onrender.com)',
      'Accept-Language': 'en'
    }
  })

  if (!res.ok) {
    throw new Error(`Nominatim ${res.status}: ${res.statusText}`)
  }

  const data = await res.json()
  if (data.length > 0) {
    return {
      lat: parseFloat(data[0].lat),
      lng: parseFloat(data[0].lon),
      display_name: data[0].display_name,
      osm_type: data[0].osm_type,
      osm_id: data[0].osm_id
    }
  }
  return null
}

const VERBOSE = args.includes('--verbose')

async function geocodeAppellation(appellation, region, country) {
  const queries = buildSearchQueries(appellation, region, country)
  const regionFallbackQuery = `${region} wine region, ${country}`

  for (const query of queries) {
    await sleep(RATE_LIMIT_MS)
    if (VERBOSE) process.stdout.write(`     trying: "${query}" ... `)
    const result = await geocodeQuery(query)
    if (result) {
      if (VERBOSE) console.log(`✓`)
      const isRegionFallback = query === regionFallbackQuery
      return { ...result, query, regionFallback: isRegionFallback }
    }
    if (VERBOSE) console.log(`✗`)
  }

  return null
}

async function main() {
  console.log(`🌍 Geocoding appellations${DRY_RUN ? ' (DRY RUN)' : ''}`)
  console.log(`   Rate limit: ${RATE_LIMIT_MS}ms per request\n`)

  // Fetch appellations that don't yet have a boundary row
  const { data: appellations, error: fetchErr } = await supabase
    .from('appellations')
    .select(`
      id,
      name,
      designation_type,
      regions!inner (
        name,
        countries!inner (
          name
        )
      )
    `)
    .order('name')

  if (fetchErr) {
    console.error('Failed to fetch appellations:', fetchErr)
    process.exit(1)
  }

  // Check which already have boundaries
  const { data: existing, error: existErr } = await supabase
    .from('geographic_boundaries')
    .select('appellation_id')
    .not('appellation_id', 'is', null)

  if (existErr) {
    console.error('Failed to fetch existing boundaries:', existErr)
    process.exit(1)
  }

  const existingIds = new Set((existing || []).map(e => e.appellation_id))
  const toGeocode = appellations.filter(a => !existingIds.has(a.id))

  console.log(`   Total appellations: ${appellations.length}`)
  console.log(`   Already geocoded: ${existingIds.size}`)
  console.log(`   To geocode: ${toGeocode.length}`)
  if (LIMIT < Infinity) console.log(`   Limit: ${LIMIT}`)
  console.log()

  const batch = toGeocode.slice(0, LIMIT)
  let success = 0
  let regionFallbacks = 0
  let failed = 0
  const failures = []

  for (let i = 0; i < batch.length; i++) {
    const a = batch[i]
    const regionName = a.regions.name
    const countryName = a.regions.countries.name
    const progress = `[${i + 1}/${batch.length}]`

    try {
      const result = await geocodeAppellation(a.name, regionName, countryName)

      if (result) {
        const fallbackTag = result.regionFallback ? ' [region fallback]' : ''
        if (result.regionFallback) regionFallbacks++
        console.log(`${progress} ✓ ${a.name} (${regionName}, ${countryName}) → ${result.lat.toFixed(4)}, ${result.lng.toFixed(4)}${fallbackTag}`)

        if (!DRY_RUN) {
          // Build the WKT point for PostGIS
          const pointWkt = `SRID=4326;POINT(${result.lng} ${result.lat})`

          const { error: insertErr } = await supabase
            .from('geographic_boundaries')
            .insert({
              appellation_id: a.id,
              centroid: pointWkt,
              boundary_confidence: 'geocoded',
              boundary_source: 'nominatim',
              boundary_source_id: result.osm_id ? `${result.osm_type}/${result.osm_id}` : null,
              boundary_updated_at: new Date().toISOString()
            })

          if (insertErr) {
            console.error(`   ⚠ Insert failed: ${insertErr.message}`)
            failed++
            failures.push({ name: a.name, region: regionName, country: countryName, error: insertErr.message })
            continue
          }
        }
        success++
      } else {
        console.log(`${progress} ✗ ${a.name} (${regionName}, ${countryName}) — no results`)
        failed++
        failures.push({ name: a.name, region: regionName, country: countryName, error: 'no geocode results' })
      }
    } catch (err) {
      console.error(`${progress} ✗ ${a.name} — ${err.message}`)
      failed++
      failures.push({ name: a.name, region: regionName, country: countryName, error: err.message })
    }
  }

  console.log(`\n━━━ Summary ━━━`)
  console.log(`✓ Geocoded:          ${success}`)
  console.log(`  ↳ Region fallback: ${regionFallbacks}`)
  console.log(`✗ Failed:            ${failed}`)
  if (failures.length > 0) {
    console.log(`\nFailed appellations:`)
    for (const f of failures) {
      console.log(`  - ${f.name} (${f.region}, ${f.country}): ${f.error}`)
    }
  }
}

main().catch(console.error)
